x-pack/plugins/security/server/session_management/session_management_service.ts

```diff
@@ -0,0 +1,160 @@
+    });
+
+    this.sessionIndex = new SessionIndex({
+      config,
+      clusterClient,
+      kibanaIndexName,
+      logger: this.logger.get('index'),
+    });
+
+    // Register task that will perform periodic session index cleanup.
+    taskManager.registerTaskDefinitions({
+      [SESSION_INDEX_CLEANUP_TASK_NAME]: {
+        title: 'Cleanup expired or invalid user sessions',
+        type: SESSION_INDEX_CLEANUP_TASK_NAME,
+        createTaskRunner: () => ({ run: () => this.sessionIndex.cleanUp() }),
+      },
+    });
+
+    return {
+      session: new Session({
+        logger: this.logger,
+        sessionCookie,
+        sessionIndex: this.sessionIndex,
+        config,
+      }),
+    };
+  }
+
+  start({ online$, taskManager }: SessionManagementServiceStartParams) {
+    this.statusSubscription = online$.subscribe(async ({ scheduleRetry }) => {
+      try {
+        await Promise.all([this.sessionIndex.initialize(), this.scheduleCleanupTask(taskManager)]);
+      } catch (err) {
+        scheduleRetry();
+      }
+    });
+  }
+
+  stop() {
+    if (this.statusSubscription !== undefined) {
+      this.statusSubscription.unsubscribe();
+      this.statusSubscription = undefined;
+    }
+  }
+
+  private async scheduleCleanupTask(taskManager: TaskManagerStartContract) {
+    let currentTask;
+    try {
+      currentTask = await taskManager.get(SESSION_INDEX_CLEANUP_TASK_NAME);
+    } catch (err) {
+      if (!SavedObjectsErrorHelpers.isNotFoundError(err)) {
+        this.logger.error(`Failed to retrieve session index cleanup task: ${err.message}`);
+        throw err;
+      }
+
+      this.logger.debug('Session index cleanup task is not scheduled yet.');
+    }
+
+    // Check if currently scheduled task is scheduled with the correct interval.
+    const cleanupInterval = `${this.config.session.cleanupInterval.asSeconds()}s`;
+    if (currentTask && currentTask.schedule?.interval !== cleanupInterval) {
+      this.logger.debug(
+        'Session index cleanup interval has changed, the cleanup task will be rescheduled.'
+      );
+
+      try {
+        await taskManager.remove(SESSION_INDEX_CLEANUP_TASK_NAME);
+      } catch (err) {
+        // We may have multiple instances of Kibana that are removing old task definition at the
+        // same time. If we get 404 here then task was removed by another instance, it's fine.
+        if (!SavedObjectsErrorHelpers.isNotFoundError(err)) {
+          this.logger.error(`Failed to remove old session index cleanup task: ${err.message}`);
+          throw err;
+        }
+      }
+    }
+
+    try {
+      await taskManager.ensureScheduled({
+        id: SESSION_INDEX_CLEANUP_TASK_NAME,
+        taskType: SESSION_INDEX_CLEANUP_TASK_NAME,
+        scope: ['security'],
+        schedule: { interval: cleanupInterval },
+        params: {},
+        state: {},
+      });
+    } catch (err) {
+      this.logger.error(`Failed to register session index cleanup task: ${err.message}`);
```

Comment:
I think we need to check for a conflict here before logging and throwing. If I'm following the task manager code correctly, then I think it'd be possible for two Kibana instances to attempt this at the same time. The first one will win, but the second one will fail, because it's using the Saved Object Repository's `create` method, without opting in to the `overwrite` functionality:

https://github.com/elastic/kibana/blob/5c770e5930f3302ded04a62d4f3f0125bb4ce7fb/x-pack/plugins/task_manager/server/task_store.ts#L174-L178

Reply:
I _believe_ it's already covered by Task Manager [here](https://github.com/elastic/kibana/blob/master/x-pack/plugins/task_manager/server/task_manager.ts#L326-L328), and as far as I understand it's the only difference between `schedule` and `ensureScheduled`.

Reply:
Oh duh, of course! I was too quick to click through, and didn't notice the exception handling there. Thanks!